pkg/auth/auth.go

```diff
@@ -147,24 +167,68 @@ func Login(ctx context.Context, systemContext *types.SystemContext, opts *LoginO
 		return nil
 	}
 	if unauthorized, ok := err.(docker.ErrUnauthorizedForCredentials); ok {
-		logrus.Debugf("error logging into %q: %v", server, unauthorized)
-		return errors.Errorf("error logging into %q: invalid username/password", server)
+		logrus.Debugf("error logging into %q: %v", key, unauthorized)
+		return errors.Errorf("error logging into %q: invalid username/password", key)
 	}
-	return errors.Wrapf(err, "authenticating creds for %q", server)
+	return errors.Wrapf(err, "authenticating creds for %q", key)
+}
+
+// parseRegistryArgument verifies the provided arg depending if we accept
+// repositories or not.
+func parseRegistryArgument(arg string, acceptRepositories bool) (key, registry string, maybeRef reference.Named, err error) {
+	if acceptRepositories {
+		key = trimScheme(arg)
+		if key != arg {
+			err = errors.New("credentials key has https[s]:// prefix")
+			return
+		}
+
+		registry = getRegistryName(key)
+		if registry == key {
+			// We cannot parse a reference from a registry, so we stop here
+			return
+		}
+
+		ref, parseErr := reference.ParseNamed(key)
+		if parseErr != nil {
+			err = errors.Wrapf(parseErr, "parse reference from %q", key)
+			return
+		}
+
+		if !reference.IsNameOnly(ref) {
+			err = errors.Errorf("reference %q contains tag or digest", ref.String())
+			return
+		}
+		maybeRef = ref
+		registry = reference.Domain(ref)
+	} else {
```

Comment:
(Absolutely non-blocking: turning this around,
```go
if !acceptRepositories {
   // short code
   return …
}
// long code
```
would require remembering less context.)

Reply:
Fixed as suggested :+1: